cdragontoolbox/hashes.py

```diff
@@ -198,45 +201,69 @@ def directory_list(self, cached=True):
+        words = list(words) # Ensure words is a list
+        format_part = "{sep}%%s" * (amount-1)
+        format_part = f"%s%%s{format_part}%s"
+        re_extract = re.compile(f"([^/_.-]+)(?=((?:[-_][^/_.-]+){{{amount-1}}})[^/]*\.[^/]+$)")
+        temp_formats = set()
+        for path in paths:
+            for m in re_extract.finditer(path):
+                match = m.group(1) + m.group(2)
+                temp_formats.add(format_part % (path[:m.start()], path[m.span()[0]+len(match):]))
+
+        formats = {fmt.replace("{sep}", sep) for fmt in temp_formats for sep in "-_"}
+
+        product = itertools.product
+        logger.info(f"substitute basename words ({amount}): {len(formats)} formats, {len(words)} words")
+        for fmt in progress_iterator(sorted(formats)):
+            self.check_iter(fmt % p for p in product(words, repeat=amount))
 
+    def _add_basename_word(self, paths, words):
+        """Add a word to all known basenames"""
+
+        words = list(words) # Ensure words is a list
         re_extract = re.compile(r'([^/_.-]+)(?=[^/]*\.[^/]+$)')
         formats = set()
-        for path in self.known.values():
+        for path in paths:
             for m in re_extract.finditer(path):
-                formats.add('%s%%s%s' % (path[:m.start()], path[m.end():]))
+                formats.update('%s%%s%s%s' % (path[:m.start()], sep, path[m.start():]) for sep in "-_")
+                formats.update('%s%s%%s%s' % (path[:m.end()], sep, path[m.end():]) for sep in "-_")
 
-        logger.info(f"substitute basename words: {len(formats)} formats, {len(words)} words")
+        unknown = self.unknown # global -> local for increased performance
+        logger.info(f"add basename word: {len(formats)} formats, {len(words)} words")
         for fmt in progress_iterator(sorted(formats)):
-            self.check_iter(fmt % s for s in words)
+            self.check_iter(fmt % w for w in words)
 
-    def _add_basename_word(self, words):
-        """Add a known word to all known basenames"""
+    def _double_substitution(self, paths, words):
+        """Replaces a word in all basenames of all given paths by two words."""
 
+        words = list(words) # Ensure words is a list
         re_extract = re.compile(r'([^/_.-]+)(?=[^/]*\.[^/]+$)')
         formats = set()
-        for path in self.known.values():
-            if "assets/characters" not in path and "vo/" not in path and "sfx/" not in path and "skins_skin" not in path: # Reduce the iteration time for game hashes
-                for m in re_extract.finditer(path):
-                    formats.update('%s%%s%s%s' % (path[:m.start()], sep, path[m.start():]) for sep in "-._")
-                    formats.update('%s%s%%s%s' % (path[:m.end()], sep, path[m.end():]) for sep in "-._")
+        for path in paths:
+            for m in re_extract.finditer(path):
+                formats.update('%s%%s%s%%s%s' % (path[:m.start()], sep, path[m.end():]) for sep in "-_")
```

Comment:
You can easily make this method generic by replacing with any required number of `%%s` then use `product(words, repeat=N)`.

Reply:
The main use for y>2 is to enlarge the set of tested paths. If you search for specific hashes you start with y=1, then try with y=2 if the first run didn't yield any result, etc.

Reply:
I'm just not sure whether I should just make a single function that offers support for all combinations of x and y, and have name links from other functions; like ``substitute_basename_words`` calling that main function with the specified amount of substitutions.

Reply:
A method that allows any combination of x and y might be a bit complicated.
If you feel confident with it, that would be great. But a `multiple_substitution()` method that offers `x = 1, y >= 2` is not hard to implement by extending `double_substitution()` and would be good enough IMO.

Reply:
It's not that difficult as ``substitute_basename_words`` already supported varying x and y, with x = y as condition. I've not split these two values. Please review the structure of the input variables.